SignalServiceKit/src/Util/OWS2FAManager.m

```diff
@@ -129,6 +140,113 @@ - (void)disable2FAWithSuccess:(nullable OWS2FASuccess)success failure:(nullable
         }];
 }
 
+
+#pragma mark - Reminders
+
+- (void)storePinCode:(nullable NSString *)pinCode
+{
+    [self.dbConnection setObject:pinCode
+                          forKey:kOWS2FAManager_PinCode
+                    inCollection:kOWS2FAManager_Collection];
+}
+
+- (nullable NSString *)pinCode
+{
+    return [self.dbConnection objectForKey:kOWS2FAManager_PinCode
+                              inCollection:kOWS2FAManager_Collection];
+}
+
+- (nullable NSDate *)lastSuccessfulReminderDate
+{
+    return [self.dbConnection dateForKey:kOWS2FAManager_LastSuccessfulReminderDateKey
+                            inCollection:kOWS2FAManager_Collection];
+}
+
+- (void)setLastSuccessfulReminderDate:(nullable NSDate *)date
+{
+    DDLogDebug(@"%@ Seting setLastSuccessfulReminderDate:%@", self.logTag, date);
+    [self.dbConnection setDate:date
+                        forKey:kOWS2FAManager_LastSuccessfulReminderDateKey
+                  inCollection:kOWS2FAManager_Collection];
+}
+
+- (BOOL)isDueForReminder
+{
+    if (!self.is2FAEnabled) {
+        return NO;
+    }
+    
+    return self.nextReminderDate.timeIntervalSinceNow < 0;
+}
+
+- (NSDate *)nextReminderDate
+{
+    NSDate *lastSuccessfulReminderDate = self.lastSuccessfulReminderDate ?: [NSDate distantPast];
+    
+    return [lastSuccessfulReminderDate dateByAddingTimeInterval:self.repetitionInterval];
+}
+
+- (NSArray<NSNumber *> *)allRepetitionIntervals
+{
+    // Keep sorted monotonically increasing.
```

Comment:
So every time they successfully guess, we "back off" and remind them less.  When they fail to guess or set a new PIN, shouldn't we reset the repetition interval to zero?

Reply:
Per the spec, I think we want to just ratchet them back one repetition interval.